src/Stratis.Bitcoin/P2P/PeerSelector.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.P2P
+{
+    /// <summary>
+    /// Contract for <see cref="PeerSelector"/>.
+    /// </summary>
+    public interface IPeerSelector
+    {
+        /// <summary>
+        /// Selects a random peer, via a selection algorithm, from the address 
+        /// manager to connect to.
+        /// </summary>
+        /// <remarks>
+        /// See the inline comments for specifics on how the peer is selected.
+        /// </remarks>
+        PeerAddress SelectPeer();
+
+        /// <summary>
+        /// Select preferred peers from the address manager for peer discovery.
+        /// </summary>
+        /// <remarks>
+        /// See the inline comments for specifics on how the peers are selected.
+        /// </remarks>
+        IEnumerable<PeerAddress> SelectPeers();
+    }
+
+    /// <summary>
+    /// Class that solely deals with providing the peer connectors a peer to connect to.
+    /// </summary>
+    public sealed class PeerSelector : IPeerSelector
+    {
+        /// <summary>
+        /// The address manager instance that holds the peer list to be queried. 
+        /// </summary>
+        private readonly IPeerAddressManager peerAddressManager;
+
+        /// <summary>
+        /// Constructor for the peer selector.
+        /// </summary>
+        /// <param name="peerAddressManager">The singleton peer address manager instance.</param>
+        public PeerSelector(IPeerAddressManager peerAddressManager)
```

Comment:
This is a bit of an anti pattern `IPeerAddressManager` depends on `IPeerSelector` and vice-versa.
Its normally not encouraged, not much point of using an interface now right?

Reply:
@dangershony you are right, I will look at improving the design in the next PR. The only thing the PeerSelector class does is to work with the peer collection. I might as well just pass the collection to it.

Reply:
yes sounds right